### PR TITLE
[KYUUBI #6278] Update DB state when the query fallback to resource manager and the batch app is terminal

### LINKED ISSUE #6278: [TASK][EASY] Update DB state when the query fallback to yarn and the batch app is terminal
### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### Search before asking

- [X] I have searched in the [issues](https://github.com/apache/kyuubi/issues?q=is%3Aissue) and found no similar issues.

### What would you like to be improved?

In scenarios with high concurrency for submitting batch sessions and querying status concurrently, especially when there are a large number of short-period batch session submissions, if some Kyuubi servers crash, it will impose significant pressure on the remaining Kyuubi servers. This is because at this time, they are unable to query the information of some batch sessions in their memory. Therefore, we should try to stack the pressure of this scenario onto YARN as much as possible. 

### How should we improve?

Therefore, I suggest that if a job reaches its final state, we should update the database as soon as possible to avoid further degradation of performance as much as possible.

### Are you willing to submit PR?

- [X] Yes. I would be willing to submit a PR with guidance from the Kyuubi community to improve.
- [ ] No. I cannot submit a PR at this time.

### PR DESCRIPTION
**Pull request by yanghua**
# :mag: Description
## Issue References 🔗
Close #6278 

This pull request fixes #

## Describe Your Solution 🔧

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

## Types of changes :bookmark:

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

#### Related Unit Tests

---

# Checklist 📝

- [ ] This patch was not authored or co-authored using [Generative Tooling](https://www.apache.org/legal/generative-tooling.html)

**Be nice. Be informative.**

### CODE REVIEW

**Comment by yanghua** [comment]
WIP.  Regardless of whether to change the code, local tests always fail. Use the community's CI to run.

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/BatchesResource.scala [hunk 1]
```diff
@@ -317,6 +317,7 @@ private[v1] class BatchesResource extends ApiRequestContext with Logging {
       sessionManager.getBatchMetadata(batchId).map { metadata =>
         if (batchV2Enabled(metadata.requestConf) ||
           OperationState.isTerminal(OperationState.withName(metadata.state)) ||
+          ApplicationState.isTerminated(ApplicationState.withName(metadata.engineState)) ||
```
Comment (turboFei) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/BatchesResource.scala:320]: engineState might be null

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/BatchesResource.scala [hunk 2]
```diff
@@ -315,8 +315,14 @@ private[v1] class BatchesResource extends ApiRequestContext with Logging {
       buildBatch(batchSession)
     }.getOrElse {
       sessionManager.getBatchMetadata(batchId).map { metadata =>
+        val isOperationTerminated = (metadata.state.nonEmpty
+          && OperationState.isTerminal(OperationState.withName(metadata.state)))
+        val isApplicationTerminated = (metadata.engineState.nonEmpty
```
Comment (turboFei) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/BatchesResource.scala:320]:
```
StringUtils.isNotBlank(metadata.engineState)
```

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/BatchesResource.scala [hunk 3]
```diff
@@ -332,6 +339,12 @@ private[v1] class BatchesResource extends ApiRequestContext with Logging {
                 Some(userName),
                 // prevent that the batch be marked as terminated if application state is NOT_FOUND
                 Some(metadata.engineOpenTime).filter(_ > 0).orElse(Some(System.currentTimeMillis)))
+              // if the batch app is terminated, update the metadata in db.
+              if (BatchJobSubmission.applicationTerminated(batchAppStatus)) {
+                sessionManager.updateMetadata(Metadata(
+                  identifier = batchId,
+                  engineState = batchAppStatus.get.state.toString))
```
Comment (turboFei) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/BatchesResource.scala:346]: should we update all the batchAppStatus into metadata?

Including appId, name, state, error.

**Review by turboFei — APPROVED** [review]
Left minor comments, cc @pan3793

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/BatchesResource.scala (hunk 3 above)
Comment (yanghua) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/BatchesResource.scala]: +1

**Review by pan3793 — APPROVED** [review]

**Pull request merged**